### ISSUE TITLE
Question: how to set resources from the snakefile

### ISSUE THREAD

**Issue by cgr71ii** [snakemake/snakemake#1866]
Hi!

There are a bunch of jobs that run in parallel, but there are some of them that I'd like to limit to a specific number of parallel jobs. I've noticed that what I want is something similar to what is asked in https://groups.google.com/g/snakemake/c/ykyWDoYBhoI?pli=1. The problem is that I'd like to define the resources, if not provided through `--resources`, from the Snakefile since there is where I have my configuration about how many jobs for the specific rules should be executed in parallel (I know that a profile or something similar would do the work, but I'd rather to don't use them since I don't want to duplicate the configuration in different files).

Example:
```python
localrules: targets, sleep_local

rule targets:
        input:expand("testSleep{localOrCluster}{sample}", localOrCluster=["Local", "Cluster"], sample=["S1","S2","S3","S4","S5","S6"])


rule sleep_local:
        output:"testSleepLocal{sample}"
        resources: qq=3
        threads: 2
        shell:"""
        sleep 10s
        touch {output}
        """

MAX_SLEEP_CLUSTER_PARALLEL_JOBS = 2

# What I want is something similar to the next line, but I don't know if it's possible
if not workflow.resources.gpu: # Check if the value was set from CLI
        workflow.resources.gpu = MAX_SLEEP_CLUSTER_PARALLEL_JOBS

rule sleep_cluster:
        output:"testSleepCluster{sample}"
        resources:
          gpu=1,
        threads: 1
        shell:"""
        sleep 10s
        touch {output}
        """
```

I'd like to be able to configure the resource `gpu` from the Snakefile in order to be able to decide the maximum parallel number of `sleep_cluster` rules which will be running in parallel. So, since the resource `gpu` from the rule `sleep_cluster` is configured to 1, if not provided it will be ignored (that is what happened when I ran the test), and if I want to run at maximum 2 `sleep_cluster` rules but I provided `-j 128`, in order to do not run all the `sleep_cluster` rules I'd need to configure that the resource `gpu` is 2 if was not provided from the CLI. Is this possible? Is there something like `workflow.resources.gpu`?

Thank you!

**Comment by cgr71ii** [comment]
Another question related to the previous one: is it possible to limit the resources in a way that is taken into account across multiple instances? I think that I want something like `--set-resource-scope gpu=global`. What I'd expect with that configuration is that if I run
```bash
snakemake --resources qq=100 gpu=1 --cores 8 -j 8 & &> ./first_execution   # First execution
sleep 1
snakemake --resources qq=100 gpu=1 --cores 8 -j 8 & &> ./second_execution  # Second execution
```

the rule `sleep_cluster` is executed in the first instance but not in the second until the first finishes. Is this possible? I'm running in a single machine these commands, so the resources are shared and I think that should be a way to tell this to snakemake.

**Comment by cgr71ii** [comment]
About the first question, I think I've found out a solution:

```python
if "gpu" not in workflow.global_resources:
    workflow.global_resources["gpu"] = MAX_SLEEP_CLUSTER_PARALLEL_JOBS
```

I've checked out the code and it seems to work as described. Of course, this might have side effects which I don't know.

About the second question I'm not sure I'll be able to find a solution since my knowledge of Snakemake is not as deep as the question requires, I think.

**Comment by pvandyken** [comment]
@cgr71ii, your solution to the first question seems reasonable to me. I can't **guarantee** it won't cause bad side effects (I don't think Snakemake *officially* supports that type of modification), but it's the way I would approach the problem, and I've done similar on other `workflow` variables.

For the second question, unfortunately that won't work. `resource-scope` only limits resources within a single snakemake instance. `scope=global` prevents snakemake from submitting multiple jobs in a cluster environment that together consume more than the resource limit.

I don't know of any obvious way of achieving your outcome: it seems more a problem of communication between processes. On the other hand, manually running multiple snakemake instances isn't exactly intended usage... Is it possible to use snakemake cluster support to orchestrate your instances? Because then `scope=global` will work perfectly.

**Comment by cgr71ii** [comment]
> I can't guarantee it won't cause bad side effects (I don't think Snakemake officially supports that type of modification), but it's the way I would approach the problem, and I've done similar on other workflow variables.

That's what I though that might be a problem, but for the moment, it's work. Of course, that doesn't mean that if I use something different which I usually don't, it might break because of this tweak. At least, is nice to know that someone else would do it the same way (and have done similar things), since it means that I'm not following the wrong direction.

By the moment I think I haven't experienced any side effect and it works the way is intended:
- https://github.com/bitextor/bitextor/blob/9d9b84befd746d5c5e8f14f3719704aff4fe981e/bitextor/Snakefile#L378
- https://github.com/bitextor/bitextor/blob/9d9b84befd746d5c5e8f14f3719704aff4fe981e/bitextor/Snakefile#L923

> I don't know of any obvious way of achieving your outcome: it seems more a problem of communication between processes.

That's exactly what it is, and following one of the FAQ answers (https://snakemake.readthedocs.io/en/stable/project_info/faq.html#i-want-to-pass-variables-between-rules-is-that-possible) I found out that this is possible. Of course, not trivial, in fact even difficult, but I've achieved a PoC that this is possible with snakemake: https://github.com/bitextor/bitextor/tree/neural_modifications

What I've done is use `PersistentStorage` and implement a basic mutex mechanism so different snakemake instance can communicate among them. Doing this, I've been able to communicate among different rules of 1 snakemake instance, but since the communication is done through a file with `PersisntentStorage`, this should be extensible to multiple snakemake instances. Maybe not directly and some changes might be needed, which I think it's not needed in my case for use this mechanism to multiple instances, but I think it's possible. The reason I think it will work even with multiple instances and not only with multiple rules instances of the same instance is because the mutex mechanism is done through `PersistentStorage`. Since all the rules are different processes, I need to communicate them using this mechanism. Since they access the same file, other snakemake instances would access the same file as well, which would lead to access the same mutex in order to ask for a GPU, and it should work.

I'll let the issue open for the moment in case that something of what I told is wrong.

Thank you, @pvandyken , for answering my doubts!
